src/main/java/gyro/google/compute/FirewallResource.java

```diff
@@ -0,0 +1,572 @@
+
+        Firewall firewall = getFirewall(client);
+
+        if (firewall == null) {
+            return false;
+        }
+
+        copyFrom(firewall);
+
+        return true;
+    }
+
+    @Override
+    public void create(GyroUI ui, State state) {
+        Compute client = createComputeClient();
+
+        try {
+            Compute.Firewalls.Insert insert = client.firewalls().insert(getProjectId(), toFirewall());
+            Operation operation = insert.execute();
+            Operation.Error error = waitForCompletion(client, operation);
+            if (error != null) {
+                throw new GyroException(error.toPrettyString());
+            }
+
+            refresh();
+        } catch (GoogleJsonResponseException je) {
+            throw new GyroException(je.getDetails().getMessage());
+        } catch (Exception ex) {
+            throw new GyroException(ex.getMessage(), ex.getCause());
+        }
+    }
+
+    @Override
+    public void update(GyroUI ui, State state, Resource current, Set<String> changedFieldNames) {
+        Compute client = createComputeClient();
+
+        Operation operation;
+        Operation.Error error;
+
+        try {
+            operation = client.firewalls().patch(getProjectId(), getName(), toFirewall()).execute();
+            error = waitForCompletion(client, operation);
+            if (error != null) {
+                throw new GyroException(error.toPrettyString());
+            }
+
+            refresh();
+        } catch (GoogleJsonResponseException je) {
+            throw new GyroException(je.getDetails().getMessage());
+        } catch (Exception ex) {
+            throw new GyroException(ex.getMessage(), ex.getCause());
+        }
+    }
+
+    @Override
+    public void delete(GyroUI ui, State state) {
+        Compute client = createComputeClient();
+
+        try {
+            Operation operation = client.firewalls().delete(getProjectId(), getName()).execute();
+            Operation.Error error = waitForCompletion(client, operation);
+            if (error != null) {
+                throw new GyroException(error.toPrettyString());
+            }
+
+        } catch (GoogleJsonResponseException je) {
+                throw new GyroException(je.getDetails().getMessage());
+        } catch (Exception ex) {
+            throw new GyroException(ex.getMessage(), ex.getCause());
+        }
+    }
+
+    @Override
+    public List<ValidationError> validate() {
+        List<ValidationError> errors = new ArrayList<>();
+
+        if (getRuleType().equals("ALLOW") && getAllowed().isEmpty()) {
+            errors.add(new ValidationError(this, "allowed", "'allowed' needs to be set when 'rule-type' set to 'ALLOW'."));
+        }
+
+        if (getRuleType().equals("DENY") && getDenied().isEmpty()) {
+            errors.add(new ValidationError(this, "denied", "'denied' needs to be set when 'rule-type' set to 'DENY'."));
+        }
+
+        if (getDirection().equals("INGRESS")) {
```

Comment:
The hope was to use the new `@ConflictsWith` annotation to get rid of this entire section of validation. See my suggestion above and let me know if that doesn't work.

Reply:
The `@ConflictsWith` can handle the conflicts arising from these validated attributes, but a value based validation will still be needed. The value of the `direction` attribute either `INGRESS` or `EGRESS` dictates which fields are allowed, which handles the conflicts as the conflicting fields belong to different values for the attribute `direction`.

We could create an annotation like `@AllowedIf(field='direction', value={'INGRESS'})` which validates a field based on what value is configured on another field. This will get rid of this whole block of validating separately.